internal/settings/providers/porkbun/provider.go

```diff
@@ -0,0 +1,295 @@
+	if err != nil {
+		return fmt.Errorf("%w: %s", errors.ErrRequestMarshal, err)
+	}
+
+	p.logger.Debug("HTTP POST createRecord: " + u.String() + ": " + string(bodyBytes))
+
+	request, err := http.NewRequestWithContext(ctx, http.MethodPost, u.String(), bytes.NewBuffer(bodyBytes))
+	if err != nil {
+		return fmt.Errorf("%w: %s", errors.ErrBadRequest, err)
+	}
+	p.setHeaders(request)
+
+	response, err := client.Do(request)
+	if err != nil {
+		return fmt.Errorf("%w: %s", errors.ErrUnsuccessfulResponse, err)
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode != http.StatusOK {
+		return fmt.Errorf("%w: %d: %s",
+			errors.ErrBadHTTPStatus, response.StatusCode, utils.BodyToSingleLine(response.Body))
+	}
+	return nil
+}
+
+func (p *provider) updateRecord(ctx context.Context, client *http.Client,
+	recordType string, ipStr string, recordID string) (err error) {
+	u := url.URL{
+		Scheme: "https",
+		Host:   "porkbun.com",
+		Path:   fmt.Sprintf("/api/json/v3/dns/edit/%s/%s", p.domain, recordID),
+	}
+	postRecordsParams := struct {
+		SecretApiKey string `json:"secretapikey"`
+		ApiKey       string `json:"apikey"`
+		Content      string `json:"content"`
+		Type         string `json:"type"`
+		TTL          string `json:"ttl"`
+		Name         string `json:"name,omitempty"`
+	}{
+		SecretApiKey: p.secretApiKey,
+		ApiKey:       p.apiKey,
+		Content:      ipStr,
+		Type:         recordType,
+		TTL:          fmt.Sprint(p.ttl),
+		Name:         p.host,
+	}
+	bodyBytes, err := json.Marshal(postRecordsParams)
+	if err != nil {
+		return fmt.Errorf("%w: %s", errors.ErrRequestMarshal, err)
+	}
+
+	p.logger.Debug("HTTP POST updateRecord: " + u.String() + ": " + string(bodyBytes))
+
+	request, err := http.NewRequestWithContext(ctx, http.MethodPost, u.String(), bytes.NewBuffer(bodyBytes))
+	if err != nil {
+		return fmt.Errorf("%w: %s", errors.ErrBadRequest, err)
+	}
+	p.setHeaders(request)
+
+	response, err := client.Do(request)
+	if err != nil {
+		return fmt.Errorf("%w: %s", errors.ErrUnsuccessfulResponse, err)
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode != http.StatusOK {
+		return fmt.Errorf("%w: %d: %s",
+			errors.ErrBadHTTPStatus, response.StatusCode, utils.BodyToSingleLine(response.Body))
+	}
+	return nil
+}
+
+func (p *provider) Update(ctx context.Context, client *http.Client, ip net.IP) (newIP net.IP, err error) {
+	recordType := constants.A
+	var ipStr string
+	if ip.To4() == nil { // IPv6
+		recordType = constants.AAAA
+		ipStr = ip.To16().String()
+	} else {
+		ipStr = ip.To4().String()
+	}
+	recordIDs, err := p.getRecords(ctx, client)
+	if err != nil {
+		return nil, err
+	}
+	if len(recordIDs) == 0 {
+		if err := p.createRecord(ctx, client, recordType, ipStr); err != nil {
+			return nil, err
+		}
+	} else {
+		for _, recordID := range recordIDs {
+			if err := p.updateRecord(ctx, client, recordType, ipStr, recordID); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return ip, nil
```

Comment:
cosmetic: reducing indentation

```suggestion
	if len(recordIDs) == 0 {
		if err := p.createRecord(ctx, client, recordType, ipStr); err != nil {
			return nil, err
		}
		return ip, nil
	}

	for _, recordID := range recordIDs {
		if err := p.updateRecord(ctx, client, recordType, ipStr, recordID); err != nil {
			return nil, err
		}
	}

	return ip, nil
```